### PR TITLE
Introduce a PUT and GET operation to the Tenant Management API (v1)

### PR DESCRIPTION
**Pull request by hwupathum**
## Purpose
Introduce new tenant management API endpoints for getting and updating tenant owner.

## Related Issue
- https://github.com/wso2/product-is/issues/21271
- https://github.com/wso2/product-is/issues/21346

## Related PRs
Merge after following PRs
- [x] https://github.com/wso2/carbon-multitenancy/pull/276
- [x] https://github.com/wso2/carbon-commons/pull/496

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java [hunk 1]
```diff
@@ -43,6 +46,12 @@ public enum ErrorMessage {
         ERROR_CODE_TENANT_LIMIT_REACHED("TM-60019",
                 "Unable to create a tenant.",
                 "Maximum number of allowed tenants have been reached."),
+        ERROR_CODE_OWNER_NOT_FOUND("TM-60020",
+                "Unable to retrieve the tenant owner.",
+                "Tenant owner cannot be found for the provided id: %s."),
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java:51]: which id ?

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java [hunk 2]
```diff
@@ -64,7 +73,11 @@ public enum ErrorMessage {
                                                         "Unable to check availability of domain.",
                 "Server encountered an error while checking for tenant domain"),
         ERROR_CODE_DELETE_TENANT_METADATA("TM-65008", "Error while deleting the tenant metadata.",
-                "Server encountered an error while deleting the tenant metadata identified by %s .");
+                "Server encountered an error while deleting the tenant metadata identified by %s ."),
+        ERROR_CODE_ERROR_RETRIEVING_OWNER("TM-65009", "Error while retrieving the owner.",
+                "Server encountered an error while retrieving the owner identified by %s ."),
+        ERROR_CODE_ERROR_UPDATING_OWNER("TM-65010", "Error while updating the owner.",
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java:79]: identified by the tenant id:

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/factory/RealmServiceOSGIServiceFactory.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.user.core.service.RealmService;
+
+/**
+ * Factory Beans serves as a factory for creating other beans within the IOC container. This factory bean is used to
+ * instantiate the RealmService inside the container.
+ */
+public class RealmServiceOSGIServiceFactory extends AbstractFactoryBean<RealmService> {
+
+    private RealmService realmService;
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/factory/RealmServiceOSGIServiceFactory.java:30]: Why not static ?

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java [hunk 4]
```diff
@@ -193,6 +199,33 @@ public List<OwnerResponse> getOwners(String tenantUniqueID) {
         }
     }
 
+    public OwnerInfoResponse getOwner(String tenantUniqueID, String ownerID, String additionalClaims) {
+
+        try {
+            Tenant tenant = TenantManagementServiceHolder.getTenantMgtService().getTenant(tenantUniqueID);
+            if (tenant.getAdminUserId() == null || !tenant.getAdminUserId().equals(ownerID)) {
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java:206]: Can tenant be null ?

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java [hunk 5]
```diff
@@ -241,6 +274,73 @@ private List<OwnerResponse> createOwnerResponse(User user) {
...
+            throw handleException(Response.Status.NOT_FOUND, TenantManagementConstants.ErrorMessage.
+                    ERROR_CODE_OWNER_NOT_FOUND, ownerID);
+        }
+        if (StringUtils.isNotBlank(ownerPutModel.getFirstname())) {
+            tenant.setAdminFirstName(ownerPutModel.getFirstname());
+        }
+        if (StringUtils.isNotBlank(ownerPutModel.getLastname())) {
+            tenant.setAdminLastName(ownerPutModel.getLastname());
+        }
+        if (StringUtils.isNotBlank(ownerPutModel.getEmail())) {
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java:291]: No email validation required ? or is it validated earlier ?

File: pom.xml [hunk 6]
```diff
@@ -813,7 +813,7 @@
         <commons.beanutils.version>1.9.4</commons.beanutils.version>
         <mavan.findbugsplugin.exclude.file>findbugs-exclude-filter.xml</mavan.findbugsplugin.exclude.file>
         <carbon.kernel.version>4.9.17</carbon.kernel.version>
-        <carbon.multitenancy.version>4.9.10</carbon.multitenancy.version>
+        <carbon.multitenancy.version>4.11.25-SNAPSHOT</carbon.multitenancy.version>
```
Comment (indeewari) [inline pom.xml:816]: Why snapshot version ?

Comment (hwupathum) [reply pom.xml:816]: Will update when the PR is merged

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java (hunk 5 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java:291]: Email validation is done inside the multitenancy repo.

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java (hunk 4 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java:206]: If tenant is not found `TenantMgtException` is thrown

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/factory/RealmServiceOSGIServiceFactory.java (hunk 3 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/factory/RealmServiceOSGIServiceFactory.java:30]: All the OSGI service factory classes in this repo has non static variables and createInstane() method

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java (hunk 1 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java]: Fixed in https://github.com/wso2/identity-api-server/pull/704/commits/cab5ddcccfde9ccfdfaf4a3f123ad205331f331f

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java (hunk 2 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java:79]: Fixed in https://github.com/wso2/identity-api-server/pull/704/commits/cab5ddcccfde9ccfdfaf4a3f123ad205331f331f

**Review by indeewari — APPROVED** [review]

**Pull request merged**